packet/packet.go

```diff
@@ -51,27 +69,38 @@ func (packet *Packet) UnmarshalBinary(data []byte) error {
 
 // MarshalBinary implements the encoding.BinaryMarshaler interface
 func (packet *Packet) MarshalBinary() ([]byte, error) {
+	// form a buffer with the packet's magic code
 	buf := bytes.NewBuffer(packet.Code)
+
+	// write the request header
 	if err := binary.Write(buf, binary.BigEndian, int32(packet.Type)); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("Error while writing packet type: %s", err)
 	}
+
+	// write the size of the packet
 	size := len(packet.Arguments) - 1 // One for each null-byte separator
 	for _, argument := range packet.Arguments {
 		size += len(argument)
 	}
-	if size < 0 {
-		size = 0
-	}
+	size = int(math.Max(0, float64(size)))
+
+	// write the size of the packet
 	if err := binary.Write(buf, binary.BigEndian, int32(size)); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("Error while writing packet length: %s", err)
 	}
-	if len(packet.Arguments) > 0 {
-		// Need special handling for last argument (don't write null byte)
-		for _, argument := range packet.Arguments[0 : len(packet.Arguments)-1] {
-			buf.Write(argument)
-			buf.Write([]byte{0})
+
+	// write all arguments provided
+	for i := 0; i < len(packet.Arguments); i++ {
+		if _, err := buf.Write(packet.Arguments[i]); err != nil {
```

Comment:
The API for `buffer.Write` specifies that the error is always nil, so you don't need to check this response